pkg/server/service/loadbalancer/wrr/wrr.go

```diff
@@ -63,76 +61,52 @@ func New(sticky *dynamic.Sticky, hc *dynamic.HealthCheck) *Balancer {
 	return balancer
 }
 
-// Len implements heap.Interface/sort.Interface.
-func (b *Balancer) Len() int { return len(b.handlers) }
-
-// Less implements heap.Interface/sort.Interface.
-func (b *Balancer) Less(i, j int) bool {
-	return b.handlers[i].deadline < b.handlers[j].deadline
-}
-
-// Swap implements heap.Interface/sort.Interface.
-func (b *Balancer) Swap(i, j int) {
-	b.handlers[i], b.handlers[j] = b.handlers[j], b.handlers[i]
-}
-
-// Push implements heap.Interface for pushing an item into the heap.
-func (b *Balancer) Push(x interface{}) {
-	h, ok := x.(*namedHandler)
-	if !ok {
-		return
-	}
-
-	b.handlers = append(b.handlers, h)
-}
-
-// Pop implements heap.Interface for popping an item from the heap.
-// It panics if b.Len() < 1.
-func (b *Balancer) Pop() interface{} {
-	h := b.handlers[len(b.handlers)-1]
-	b.handlers = b.handlers[0 : len(b.handlers)-1]
-	return h
-}
-
 // SetStatus sets on the balancer that its given child is now of the given
-// status. balancerName is only needed for logging purposes.
-func (b *Balancer) SetStatus(ctx context.Context, childName string, up bool) {
-	b.mutex.Lock()
-	defer b.mutex.Unlock()
+// status.
+func (b *Balancer) SetStatus(ctx context.Context, childName string, healthy bool) {
+	log.FromContext(ctx).Debugf("Setting status of %s to %v", childName, statusAsStr(healthy))
 
-	upBefore := len(b.status) > 0
-
-	status := "DOWN"
-	if up {
-		status = "UP"
-	}
-	log.FromContext(ctx).Debugf("Setting status of %s to %v", childName, status)
-	if up {
-		b.status[childName] = struct{}{}
-	} else {
-		delete(b.status, childName)
+	b.mutex.Lock()
+	nh := b.handlersByName[childName]
+	if nh == nil {
+		b.mutex.Unlock()
+		return
 	}
 
-	upAfter := len(b.status) > 0
-	status = "DOWN"
-	if upAfter {
-		status = "UP"
+	healthyBefore := b.healthyCount > 0
+	if nh.healthy != healthy {
+		nh.healthy = healthy
+		if healthy {
+			b.healthyCount++
+			b.enabledHandlers.push(nh)
+		} else {
+			b.healthyCount--
```

Comment:
We cannot call `pop` here because the unhealthy node is not guaranteed to be at the top of the heap. We do not know its location in the heap so we just mark it as unhealthy, it is removed from the heap in `acquireHandler` when it is returned by `pop`.